pkg/server/commands/list_objects.go

```diff
@@ -95,76 +113,121 @@ func (q *ListObjectsQuery) evaluate(
-
-			span.SetAttributes(attribute.Bool(listObjectsOptimizedKey, true))
+		connectedObjectsResChan := make(chan *ConnectedObjectsResult, 1)
+		var objectsFound = new(uint32)
 
-			connectObjCmd := &ConnectedObjectsCommand{
-				Datastore:        q.Datastore,
-				ResolveNodeLimit: q.ResolveNodeLimit,
-				Limit:            maxResults,
-			}
+		connectedObjectsCmd := &ConnectedObjectsCommand{
+			Datastore:        q.Datastore,
+			Typesystem:       typesys,
+			ResolveNodeLimit: q.ResolveNodeLimit,
+			Limit:            maxResults,
+		}
 
-			err = connectObjCmd.StreamedConnectedObjects(ctx, &ConnectedObjectsRequest{
+		go func() {
+			err = connectedObjectsCmd.StreamedConnectedObjects(ctx, &ConnectedObjectsRequest{
 				StoreID:          req.GetStoreId(),
 				Typesystem:       typesys,
 				ObjectType:       targetObjectType,
 				Relation:         targetRelation,
-				User:             targetUserRef,
+				User:             sourceUserRef,
 				ContextualTuples: req.GetContextualTuples().GetTupleKeys(),
-			}, resultsChan)
+			}, connectedObjectsResChan)
+			if err != nil {
+				resultsChan <- ListObjectsResult{Err: err}
+			}
+
+			close(connectedObjectsResChan)
+		}()
+
+		checkResolver := graph.NewLocalChecker(
+			q.Datastore,
+			q.CheckConcurrencyLimit,
+		)
+
+		concurrencyLimiterCh := make(chan struct{}, maximumConcurrentChecks)
+
+		wg := sync.WaitGroup{}
+
+		for res := range connectedObjectsResChan {
+
+			if res.ResultStatus == NoFurtherEvalStatus && atomic.AddUint32(objectsFound, 1) <= maxResults {
+				noFurtherEvalRequiredCounter.Inc()
+				resultsChan <- ListObjectsResult{ObjectID: res.Object}
+				continue
+			}
+
+			furtherEvalRequiredCounter.Inc()
+
+			wg.Add(1)
+			go func(res *ConnectedObjectsResult) {
+				defer func() {
+					<-concurrencyLimiterCh
+					wg.Done()
+				}()
+
+				concurrencyLimiterCh <- struct{}{}
+
+				resp, err := checkResolver.ResolveCheck(ctx, &graph.ResolveCheckRequest{
+					StoreID:              req.GetStoreId(),
+					AuthorizationModelID: req.GetAuthorizationModelId(),
+					TupleKey:             tuple.NewTupleKey(res.Object, req.GetRelation(), req.GetUser()),
+					ContextualTuples:     req.GetContextualTuples().GetTupleKeys(),
+					ResolutionMetadata: &graph.ResolutionMetadata{
+						Depth: q.ResolveNodeLimit,
+					},
+				})
+				if err != nil {
+					resultsChan <- ListObjectsResult{Err: err}
+					return // is this the right thing here or should we try to find all of them we can up to some limit?
+				}
+
+				if resp.Allowed && atomic.AddUint32(objectsFound, 1) <= maxResults {
+					resultsChan <- ListObjectsResult{ObjectID: res.Object}
+				}
+			}(res)
+		}
+
+		wg.Wait()
+
+		close(resultsChan)
+	}
+
+	containsIntersection, _ := typesys.RelationInvolvesIntersection(targetObjectType, targetRelation)
+	containsExclusion, _ := typesys.RelationInvolvesExclusion(targetObjectType, targetRelation)
+
+	if (containsIntersection || containsExclusion) && !q.OptimizeIntersectionExclusion {
```

Comment:
if the flag is off, why do we need to check if the model has `and` or `but not`?

Reply:
Because if the model contains intersection OR exclusion then we must fall back to concurrent Checks. Otherwise the default handler (e.g. the handler that evaluates using ConnectedObjects) is used. So this covers that case properly.

Reply:
Ah makes sense